KEP: [Cluster-Lifecycle] Install Addons via kubeadm (#1308)
kubernetes/enhancements: keps/sig-cluster-lifecycle/addons/20191013-install-addons-via-kubeadm.md

---
title: Install Addons via kubeadm
authors:
  - "@stealthybox"
  - "@fabriziopandini"
owning-sig: sig-cluster-lifecycle
participating-sigs:
reviewers:
  - TBD
  - "@dholbach"
  - "@fabriziopandini"
  - "@justinsb"
  - "@neolit123"
  - "@rosti"
approvers:
  - TBD
  - "@timothysc"
editor: TBD
creation-date: 2019-10-13
last-updated: 2019-10-13
status: provisional #|implementable|implemented|deferred|rejected|withdrawn|replaced
see-also:
  - "/keps/sig-cluster-lifecycle/addons/0035-20190128-addons-via-operators.md"
  - "/keps/sig-cluster-lifecycle/wgs/0032-create-a-k8s-io-component-repo.md"
replaces:
superseded-by:
---

# Install Addons via kubeadm

## Table of Contents

> <!-- [Tool for generating](https://github.com/ekalinin/github-markdown-toc) -->

Author: I left this comment in here since many folks copy and paste KEP's.
(Generating the table of contents is a big time saver.)

**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://github.com/kubernetes/enhancements/issues
[kubernetes/kubernetes]: https://github.com/kubernetes/kubernetes
[kubernetes/website]: https://github.com/kubernetes/website -->

## Summary

<!-- The `Summary` section is incredibly important for producing high quality user-focused documentation such as release notes or a development roadmap.
It should be possible to collect this information before implementation begins in order to avoid requiring implementors to split their attention between writing release notes and implementing the feature itself.
KEP editors, SIG Docs, and SIG PM should help to ensure that the tone and content of the `Summary` section is useful for a wide audience.

A good summary is probably at least a paragraph in length. -->

> The addons subproject hosted by SIG Cluster Lifecycle has been working on implementation details regarding addon installation, user-experience, maintenance, and operation.

Reviewer: Suggest introducing the term "addon" briefly here, so someone who hasn't heard the back-story can get off on the right foot.
Probably link to https://github.com/kubernetes/enhancements/blob/master/keps/sig-cluster-lifecycle/addons/0035-20190128-addons-via-operators.md

Reviewer 2: If https://github.com/kubernetes/website/pull/18156 lands it'll be possible to hyperlink to the entry about add-ons in the Kubernetes glossary too.

Reviewer 2: It landed, so you can link to https://kubernetes.io/docs/reference/glossary/?tool=true#term-addons

Author: Thanks for adding the link here @dholbach

The [addon-installer library POC](https://github.com/kubernetes-sigs/addon-operators/pull/25) functions as a small, vendorable implementation for applying addons to the cluster using kubectl and Kustomize. The `AddonInstallerConfiguration` API kind is intended to be a user-facing ComponentConfig.
This KEP proposes an additional phase to kubeadm that invokes the addon-installer library.

## Motivation

<!-- This section is for explicitly listing the motivation, goals and non-goals of this KEP.
Describe why the change is important and the benefits to users.
The motivation section can optionally provide links to [experience reports][] to demonstrate the interest in a KEP within the wider Kubernetes community. -->

Kubeadm currently installs two "core" addons into the cluster once the control-plane is running:  kube-proxy and CoreDNS.
Optionally, it also supports installing kube-dns in place of CoreDNS.
Installation of these addons occurs via application of manifests for which the strings are built into and versioned with the kubeadm binary.
It is not currently possible to disable installation of these addons or modify them with a formal kubeadm init.

> Upgrades are accomplished with modifications to kubeadm's upgrade logic, often by introspecting which GVK's exist in the current cluster. 

Author: Another example is the DNS-change post-upgrade logic.
It chooses Deployment names to wait-for and delete based on the ClusterConfiguration, then blocks on Ready for the Delete:
https://github.com/kubernetes/kubernetes/blob/af231b09e41ae83ba4af068cc41f1f5d7340b354/cmd/kubeadm/app/phases/upgrade/postupgrade.go#L111-L140

> Upgrades are accomplished with modifications to kubeadm's upgrade logic, often by introspecting which GVK's exist in the current cluster. 

Author: I do mean GVK's since it's not just ConfigMaps:

https://github.com/kubernetes/enhancements/pull/1308#discussion_r353257068
> Another example is the DNS-change post-upgrade logic.
> It chooses Deployment names to wait-for and delete based on the ClusterConfiguration, then blocks on Ready for the Delete:
> [kubernetes/kubernetes:cmd/kubeadm/app/phases/upgrade/postupgrade.go@`af231b0`#L111-L140](https://github.com/kubernetes/kubernetes/blob/af231b09e41ae83ba4af068cc41f1f5d7340b354/cmd/kubeadm/app/phases/upgrade/postupgrade.go#L111-L140)

Kubeadm installs these "core" addons leaving an almost functioning cluster available for the user, but a CNI implementation is not installed. That is left to the cluster operator.

> Maintaining these manifests and this logic with kubeadm has to this point has allowed users to recieve a useful, default setup.

Reviewer 3: ```suggestion
Maintaining these manifests and this logic with kubeadm has to this point has allowed users to receive a useful, default setup.
```

As the ecosystem grows, this is proving to now be too inflexible for users and vendors.

Use of ClusterAPI means less operators will be manually mutating clusters before they are expected to function.

> Kubeadm does not need to maintain CNI installation, but users should be able to use kubeadm to produce a functioning cluster, complete with with operational CNI. 

Reviewer: Can we say something about creating a level playing-field for 3rd-party addons beside the set that Kubernetes/kubeadm maintainers have decided to include?

Author: @dholbach added a mention -- thanks

> ### Goals

Author: :+1:

<!-- List the specific goals of the KEP.
How will we know that this has succeeded? -->

> Integrating the addon-installer library and API with kubeadm will allow users to opt out of installing kube-proxy or CoreDNS/kube-dns.

Author: :+1:

Some reasons for this may be:
- The user does not desire pod networking and/or service-discovery in their cluster.
- CNI implementations such as kube-router and Cilium can actually replace kube-proxy.

The "core" addons can be referenced and installed from a central, extensible, shared implementation.

> In the case of CoreDNS, this may even make sense to be managed by the upstream project instead of maintaining copies of manifests in kubernetes/kubernetes.

Author: This feels out of scope for the first iteration of an alpha feature.
CoreDNS team members seem pretty committed to the operator approach.

We can maintain full ownership of these components until a formal hand-off (if any) is made.
Would you like me to add this to the KEP?

This integration will also allow users to extend these "core" addons using their own kustomize driven patches.
Patches are expected to be supplied via the API, filesystem, or git with additional potential for OCI images.

> This will also enable users to specify additional addons to install such as CNI implementations, drivers, webhook-servers, or other API's and operators.

Reviewer 3: ```suggestion
This will also enable users to specify additional addons to install such as CNI implementations, drivers, webhook-servers, or other APIs and operators.
```

Simple upgrades to newer manifests for a given addon should be supported.
Removal should also be supported.

> The mechanisms and API's for doing this are independently factored from kubeadm meaning other installers such as kops and eksctl as well as cloud and vendor solutions may implement their own defaulting logic and user-interfaces.

Reviewer 3: ```suggestion
The mechanisms and APIs for doing this are independently factored from kubeadm meaning other installers such as kops and eksctl as well as cloud and vendor solutions may implement their own defaulting logic and user-interfaces.
```

### Non-Goals

Addons are currently expected to be applied in order with no success condition beyond API validity.
No dependency system or DAG is currently proposed or in-scope.

> Discussions of how addons may validate ComponentConfigs and other files and how the AddonInstaller API can enable that are welcome. There are existing efforts in this area already (currently championed by @rosti) that may need to be cross-referenced.

Author: > It may be more appropriate to exec `kustomize` directly or vendor it in for manifest generation so that dryRun and validation of the kustomize addons will work in the absence of a working kube-apiserver. (This is important before `kubeadm init`)

I agree.
This adds technical complexity, but we can work toward it.

> Usable operators for addons such as the "CoreDNS Operator" are for consideration regarding the motivation, UX, and vision of this KEP's features. However, implementation, packaging, and delivery of these operators is out of scope for the technical implementation of the AddonInstaller in kubeadm.

Reviewer: I found it hard to parse "for consideration regarding the".  Did you mean something like "may be considered in future" ?

Author: @dholbach reworded this :+1:

> Providing an "uninstall" signal, hook, or mechanism for addons is currently out of scope but is open for discussion.

Author: It is a pre-existing issue with `kubeadm reset`.

ex: When `kubeadm reset` is run:
- A `kube-proxy` addon receives no notice that it should cleanup IPTables/IPVS.
- A `calico` addon receives no notice that it should cleanup the CNI config.

## Proposal

> An alpha feature gate "`AddonInstaller`" will be added to kubeadm; it will default to false may be user-enabled using the exisiting flag and `ClusterConfiguration`.

Reviewer 3: ```suggestion
An alpha feature gate "`AddonInstaller`" will be added to kubeadm; it will default to false may be user-enabled using the existing flag and `ClusterConfiguration`.
```

When this feature gate is enabled, the current `kube-proxy` and `coredns` phases under `init.addon` will noop.

> A new phase called `installer` will activate creating a sensible logger and runtime for `sigs.k8s.io/addon-operators/installer` libary.

Reviewer 3: ```suggestion
A new phase called `installer` will activate creating a sensible logger and runtime for `sigs.k8s.io/addon-operators/installer` library.
```

This phase is expected to operate in an idempotent manner, run properly during `kubeadm init` and `kubeadm upgrade`, and produce helpful output in a dry-run.

> An `AddonInstallerConfiguration` will be defaulted but may be overriden by the user.

Reviewer 3: ```suggestion
An `AddonInstallerConfiguration` will be defaulted but may be overridden by the user.
```

It is interpreted to be the authoritative declaration of addons in the cluster.

The phase can roughly operate in this manner:
- Parse the kubeadm global `--dry-run` flag.
- Determine if a KUBECONFIG is available for the cluster.
- Load an `AddonInstallerConfiguration` from the multi-YAML-document kubeadm `--config`.
  If missing, default to one that installs the version appropriate manifests for kube-proxy and the user-indicated DNS solution.

> - If possible, pull the previous `AddonInstallerConfiguration` from the cluster ConfigMap.

Reviewer 4: The tricky question here is what we do when?
Do we always take locally supplied `AddonInstallerConfiguration` (and possibly end with something like `kubeadm upgrade apply --config`)?
Do we default if we can't find such a config everywhere (neither file, nor cluster)?
I assume, that different kubeadm sub-commands will operate slightly differently here.

Author: @rosti:
Yes, `kubeadm upgrade apply --config` is the current suggested mechanism.
I implemented it in the POC so that it would be possible for kubeadm users to specify different addon versions on upgrade.
When a user provides a config, their intent is authoritative. We can write heuristic warnings, but we should be careful of doing anything automatically for a user WRT top-level config declarations.
I agree that the Config can be defaulted when none is present locally.

My instinct says the in-cluster config is not the authoritative declaration of user intent, even when the user provides no config.
If no config is provided on upgrade, kubeadm can generate and default it which allows for the existing U/X where kubeadm can upgrade kube-proxy/CoreDNS/kubedns.

Say the user were to provide an empty AddonInstallerConfig -- in this case, you wouldn't default it. The existing in-cluster config is then used to prune the addons not present in the empty user-provided config.

@fabriziopandini 
If a new API is introduced, I would put that in the release notes and expect users to update any user-defined AddonInstallerConfigs.
The U/X of this is a generic and unsolved ComponentConfig issue that I am willing to continue helping with.
ex: a high-level solution for reading a `v1alpha1` API from in the cluster using a newer kubeadm would be to:
- `1.`  convert to a hub version such as `Internal` or `v1alpha2`
- `2.`  use a api-specific conversion hook provided by:
  - `a.`  imported api-machinery
  - `b.`  conversion exec hook, similar to conversion-webhooks, using:
    -   ` i.`  binary in path
    -   `ii.`  binary from OCI image executed though CRI

Providing a generic solution for this is out of scope for this Addons KEP.

- Create an installer library runtime with: dryRun, kubeConfig, oldInstallCfg, and newInstallCfg.
- Prune addons from the cluster if names from the previous configuration are missing in the new configuration.
  The library should remove addons serially in reverse order.
- Apply all addons in the new configuration.
  The library should respect the declared, serial order.

> - Apply the new `AddonInstallerConfiguration` to the cluster ConfigMap.

Author: In the POC, I chose to create its own ConfigMap.
IIRC, this is similar to the kubeletConfig.

Persisting this declaration into the cluster on-success allows for declarative upgrade/remove operations of the kubeadm managed addons.

> Removing resources from the kuberneted API currently does not propogate any "delete reason" to processes in Pods.

Reviewer 3: ```suggestion
Removing resources from the kubernetes API currently does not propagate any "delete reason" to processes in Pods.
```

Uninstall Support Proposal A [not currently in scope]:
- An uninstall kustomize dir can be optionally provided with an addon (along with `name` + `ref`) when applied.
- When the addon is removed, either explicitly or as a consequence of prune via declaration, these manifests could be applied.
- Any uninstall Jobs can be polled for completion.
- If successful:
  - All addon manifests are removed.
  - All uninstall manifests are removed.

### User Stories

Users may enable this feature in `kubeadm init` and `kubeadm upgrade` using the `AddonInstaller` feature gate:
```shell
kubeadm init --feature-gates AddonInstaller=true
kubeadm init --config
```
```yaml
apiVersion: kubeadm.k8s.io/v1beta2
kind: InitConfiguration
---
apiVersion: kubeadm.k8s.io/v1beta2
kind: ClusterConfiguration
featureGates:
  AddonInstaller: true
```

Users may invoke the phase similarly:
```shell
kubeadm init phase addon installer --feature-gates AddonInstaller=true
kubeadm init phase addon installer --config
```

> The feature gate is still expected to be enabled/true when invoking the phase directly.

Author: I don't see a need for the default addons logic to be phase specific.
They can be inhibited by the user passing `--config`.

This comment about setting the feature-gate is about requiring the user to set the feature-gate when invoking the `addon installer` phase. Sorry if that was unclear.

> The feature gate is still expected to be enabled/true when invoking the phase directly.

Author: I made the verbiage more clear.

Users may print the default `AddonInstallerConfiguration` using `kubeadm config`:
```shell
kubeadm config print init-defaults --feature-gates AddonInstaller=true
# this requires adding a feature-gates flag to this command
```
TODO: Discuss this alternative:
```shell
kubeadm config print addon-defaults
```
Similar U/X can be expected for `kubeadm config migrate` and `kubeadm config view`

The user may extend the default kubeadm config using the `AddonInstallerConfiguration` kind.
The API details are still being finalized. Usage may look like this:
```yaml
---
apiVersion: addons.config.k8s.io/v1alpha1
kind: AddonInstallerConfiguration
addons:
- name: kube-proxy

>   ref: github.com/kubernetes/kubernetes//cluster/addons/kustomize/kube-proxy/?ref=v1.17.0

Author: Yes!
In my eyes, this a huge benefit in the freedom of the user.
The choice to not install a core addon and replace it with something else, use a privately patched version, use a fork, etc. is a big motivator for adopting this interface.

kube-proxy is a great example of something people want more control over.

For the user who wants things to just work out-of-the-box, the kubeadm defaults will take care of them from upgrade to upgrade.

- name: coredns-operator
  ref: github.com/kubernetes-sigs/addon-operators//kustomize/coredns-operator/?ref=v0.1.0

> - name: weavenet

Reviewer 5: I think it would be nice to expose a type here and some semantics around deletion.

I bring this up because I think we need more structure here to handle the operator case. An operator will likely bring along a CRD, CR, and Operator (eg rbac/deployment). I think we should ensure that the addon installer either meets or has the flexibility to implement this part of the lifecycle. 

When an addon is removed from the this list, we ought to provide a graceful cleanup. For operators this is typically done by managing a finalizer on the CR. If we treat them all as a single package then the CR/CRD/Operator will all deleted at the same time. In that case, the CR/CRD/workload may never be removed because the operator that set the finalizer has been removed.

A sketch of this structure could look like:
```yaml
- name: coredns
  refs:
  - type: apiextensions.k8s.io/v1,Kind=CustomResourceDefinition
    url: github.com/kubernetes-sigs/addon-operators/coredns-operator/config/crd
  - type: addons.x-k8s.io/v1alpha1,Kind=CoreDNS
    manifest: -|
      apiVersion: addons.k8s.io/v1alpha1
      kind: CoreDNS
      metadata:
        name: default
        namespace: kube-system
      spec:
        channel: stable
        patches:
        - apiVersion: apps/v1
          kind: Deployment
          metadata:
            name: coredns
            namespace: kube-system
            labels:
              foo: bar
  - type: kustomize # or addons.k8s.io/Operator?
    url: github.com/kubernetes-sigs/addon-operators/kustomize/coredns-operator
```

We could delete everything but the operator, wait till the objects are fully cleaned up, then delete the operator.

  ref: oci+kustomize://weaveworks/kustomize-weavenet:v2.5.2
    # kustomize OCI backend currently unimplemented
    # POC packaging:  https://github.com/ecordell/kpg
- name: my-addon1
  ref: ../my-local-dir/addon1/
```

> It may be in scope for kubeadm to internally patch the `AddonInstallerConfiguration` with the contents of user-supplied ComponentConfigs such as `KubeProxyConfiguration`.

Reviewer 6: Apart from component config customization, there are also other well-known users requirements for addons customizations like e.g.
- changing the image repository name (e.g for air-gapped environments)
- changing the image name (e.g for enterprises using internally built images)

How do you envision kubeadm/the user being able to do this in a generic way (that works with any addon)?

Reviewer 6: @rosti if I got this right, it seems that kubeadm could not get rid of the dependency for the component config for kube-proxy + there will be an additional dependency on the addon-installer component config; 
Do you think that this can impact on the scope of current kube-proxy rework activities (or require further activities)?

Reviewer 6: Also, defining the logistic about how to manage the kubeadm manage component configuration is important (e.g. for upgrades)

> Various patchList fields may need to be added to the `Addon` type of the library's API to support this specific U/X as well as more lightweight user-extension of addons.

Reviewer 6: @rosti  proposal about component config includes a set of ideas about how to manage the lifecycle of user-provided component config patches, e.g. for validation and upgrades.
Are you planning to follow the same approach? It is worth to document it (mostly for the expected impacts on the users)

The AddonInstaller feature composes with kubeadm's global `--dry-run` flag:
- on `init`, the intended addons (and perhaps manifests) can simply be printed.

> - on `upgrade diff` and `upgrade apply`, the expected difference relative to the cluster can be output.

Author: Thanks for mentioning this.
I think showing a diff between the declared `AddonInstallerConfig`'s is the appropriate way to communicate the top-level changes for addons.
I'll add a note to the KEP :+1:

  `kubectl` already does this currently.

The difference in behavior between a dryRun init and upgrade is caused by the presence of a working KUBECONFIG and control-plane.

Air-gap support may be accomplished through the mirroring of git repositories or images containing the kustomize dir.
Mirroring git repositories, cloning git repositories to a local filesystem, and hosting OCI images would seem to all be a reasonable capabilities for a user in an air-gapped environment. The API allows a user to reference addon sources for all of these use-cases.

#### Links

@fabriziopandini began a design doc for kubeadm integration:
- https://docs.google.com/document/d/1FZA873LBXf-aG4UULOk96MhT8xm_nPHq5Vu0z61S8pk

More generic User Stories have been collected by @dholbach:

> - doc: https://docs.google.com/document/d/17NH4xcFdeh4NVcIBjMQNK2P27j9xrsm4sCFTXdZqopg

Reviewer 3: I think we should just keep the spreadsheet - the other doc was just a list to get the conversation started.

- sheet: https://docs.google.com/spreadsheets/d/1Np0aOQYyiRqRQYOUoMcUwP_2Tt9b39HUwJXQIgW9rYA

### Implementation Details/Notes/Constraints

We intend for the installer library and API to live in `sigs.k8s.io/addon-operators/installer`.
Core kustomize addons and images may be hosted in a number of git repositories or image registries.

Internally, the addon-installer library execs `kubectl -k`.

> This introduces a binary and PATH dependency for `kubeadm` when using this feature.

Reviewer 6: Is this a dependency that applies only to control plane nodes?

Author: Yes, it's only required for nodes expected to perform `init`/`upgrade`.

Furthermore, certain sensible usage of this feature may depend on `git` by proxy of kustomize.
While `kubectl` may commonly be available in the cluster-installation environment, `git` may not be available.
ex: The current kinder images do not have `git`.

> It may be more appropriate to exec `kustomize` directly or vendor it in for manifest generation so that dryRun and validation of the kustomize addons will work in the absence of a working kube-apiserver. (This is important before `kubeadm init`)

Reviewer 6: Agree + As per comment in the google Doc, I'm not super happy by the shelling out approach, because it introduces also potential incompatibilities due to kubectl version skew.
Are you planning to implement preflight checks to prevent version skew problems?

Author: I can add a version-specific pre-flight check that is invoke-able from the library since this is an internal detail.
There is already an in-library LookPath check for `kubectl`.

If we change the internals, there will be no external dependency, but the versioning burden will move into the `go mod` of the library, and propagate into the kubeadm and kops dependency chain. This will incur more maintenance, but may be the right approach.

The kubeconfig may be passed to the library as a filepath.

> This works for kubeadm and means less work when exec'ing `kubectl` for the apply logic.

Reviewer 4: \* `executing`

Should we also support a formal internal type for this?

A more mature addon-installer library may:
- Vendor kustomize for manifest generation.
- Use an apiClient type for applying the manifests.
  ( This may be compatible with some shareable client type supporting dryRun? )
- Have an go-native git clone implementation

Kustomize currently supports constructing manifests from local directories and git.
OCI image support in kustomize is currently unimplemented.
A POC of some of this mechanism is implemented here: https://github.com/ecordell/kpg
Implementation of this feature could be delegated to another KEP.

> ### Risks and Mitigations

Reviewer 6: I think it is missing a risk about impacts on kubeadm users / third part tools using kubeadm

<!-- What are the risks of this proposal and how do we mitigate.
Think broadly.
For example, consider both security and how this will impact the larger kubernetes ecosystem.

How will security be reviewed and by whom?
How will UX be reviewed and by whom?

Consider including folks that also work outside the SIG or subproject. -->

> The core of this proposal is that it expands on the curret packaging of addons within the kubeadm binary.

Reviewer 3: ```suggestion
The core of this proposal is that it expands on the current packaging of addons within the kubeadm binary.
```

This binary is currently signed, released, and distributed by kubernetes release machinery.

> Since this proposal necessitiates new artifacts and distribution channels for default/"core" components, SIG Release will ultimiately be interested in providing methods of verification and integrity of those artifacts for users.

Reviewer 3: ```suggestion
Since this proposal necessitates new artifacts and distribution channels for default/"core" components, SIG Release will ultimately be interested in providing methods of verification and integrity of those artifacts for users.
```

Git is a strong tool for versioning manifests and patches but may be questionable for asserting source provenance without instrumenting non-default behavior.
This attack surface may become more evident when mirroring in an airgap environment.
When fetching/cloning from GitHub the transport will either be verified by HTTPS or an SSH host key.

Distributing kustomize dirs via signed OCI images may be a solution for this.

> Proposing additional support for raw manifest lists (with or without kustomization) may also be effective for certain use-cases when sourced over HTTPS from cloud-storage such as GitHub-raw, S3, GCS, or an internal solution like NGINX or Minio.

Reviewer 4: We might want to consider some form of signing stuff here. The prospect of fetching a raw manifest from some bucket somewhere (albeit the HTTPS) and applying it over the cluster with an administrative account is kind of scary to me.

---
title: Install Addons via kubeadm
authors:
  - "@stealthybox"
  - "@fabriziopandini"
owning-sig: sig-cluster-lifecycle
participating-sigs:
reviewers:
  - TBD
  - "@dholbach"
  - "@fabriziopandini"
  - "@justinsb"
  - "@neolit123"
  - "@rosti"
approvers:
  - TBD
  - "@timothysc"
editor: TBD
creation-date: 2019-10-13
last-updated: 2019-10-13
status: provisional #|implementable|implemented|deferred|rejected|withdrawn|replaced
see-also:
  - "/keps/sig-cluster-lifecycle/addons/0035-20190128-addons-via-operators.md"
  - "/keps/sig-cluster-lifecycle/wgs/0032-create-a-k8s-io-component-repo.md"
replaces:
superseded-by:
---

# Install Addons via kubeadm

## Table of Contents
<!-- [Tool for generating](https://github.com/ekalinin/github-markdown-toc) -->

> - [Install Addons via kubeadm](#install-addons-via-kubeadm)

Reviewer 7: k8s-ci-bot says that an error occured at [here](https://github.com/kubernetes/enhancements/pull/1308#issuecomment-564609993) and says that `... 20191013-install-addons-via-kubeadm.md: missing opening TOC tag` in [detailed log](https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/enhancements/1308/pull-enhancements-verify/1204792352162451456#1:build-log.txt%3A39)
It was occured on `hack/verify-toc.sh`.
Maybe, aren't `<!-- toc -->` and `<!-- /toc -->`  needed here?

Author: Thanks for the guidance on this :+1:
